executor/include/execution/trait/is_executor.hpp

```diff
@@ -5,28 +5,31 @@
 #pragma once
 
 #include <execution/executor/base_executor.hpp>
+#include <execution/trait/common_traits.hpp>
 #include <experimental/type_traits>
 #include <type_traits>
 
 namespace execution {
 
-template <class Executor, typename Function = void (*)(), typename = void>
+template <class Executor, typename = void>
 struct is_executor : std::false_type {};
 
 namespace detail {
 const auto noop = [] {};
 }
 
-template <typename Executor, typename Function>
+template <typename Executor>
 struct is_executor<
```

Comment:
I think I asked you before and haven't seen an answer yet, can you describe in words what defines an executor? I read the following from your code:
1. It needs to be copy constructible. - I'm not sure if this is semantically important, given that I've now seen you accepting universal references for executor types. I can see practical use behind making sure this is true.
2. It needs to have a method named `execute`, which takes an arbitrary functor as argument and returns void.
3. It defines an equivalence operator. - This one I don't get. I don't understand why this is a mandatory property of an executor

Other things I believe are important but not specified here:
a. Executors all derive from the same `executor` base class. This enforces that they all have the 4 properties defined and that query methods of each property exists. 

Ultimately I consider 2. and a. to be important. The rest is meh. Hence my question, can you describe in words what makes an executor, an executor? What are the defining characteristics.

Reply:
Sorry I missed your question.
From Section 2.2.9 of http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0443r13.html
```cpp
    template<class E, class F>
      concept executor-of-impl =
        invocable<remove_cvref_t<F>&> &&
        constructible_from<remove_cvref_t<F>, F> &&
        move_constructible<remove_cvref_t<F>> &&
        copy_constructible<E> &&
        is_nothrow_copy_constructible_v<E> &&
        equality_comparable<E> &&
        requires(const E& e, F&& f) {
          execution::execute(e, (F&&)f);
        };
```
This is the currently proposed concept which an executor must satisfy.
A simpler and clearer wording of this exists here open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1660r0.pdf in Section 5.1.

Coming to your points:
1 and 2, I think you already have put forward its use.
Regarding 3, being able to tell if two executors are the same or not might seem trivial right now as they will have the same type, but there is no requirement that executor properties can't be stateful and if they are stateful then there should be a way to be able to distinguish two executors since comparing types will not be enough.

As of use cases, I am not completely aware of it currently. But I am trying my best to follow and satisfy all the requirements mentioned in the proposal.

Also, regarding `base_executor` there is not needed and I have removed it in an upcoming PR. There is no requirement that the executors satisfy all those 4 properties. An executor can implement any set of properties.
Adding on that out of the 4 to have been removed as they are no longer part of the proposal namely, `Interface` and `Cardinality`.

Reply:
Golden answer. I'll need to revise some of my comments to your 0.2 proposal.